src/librustc_codegen_llvm/type_of.rs

```diff
@@ -252,12 +261,18 @@ impl<'tcx> LayoutLlvmExt<'tcx> for TyLayout<'tcx> {
                 return llty;
             }
             let llty = match self.ty.sty {
+                ty::TyRef(_, ty, _) if cx.type_is_freeze(ty) => {
+                    cx.layout_of(ty).llvm_type(cx)
+                        .to_const_ptr(cx)
```

Comment:
So how does a `&mut T` -> `&T` borrow work? Is a `addrspacecast` instruction generated? If so, why choose const/mutable *ahead* of time? Is this solely an optimization?

In other words, would it be possible to make a *less optimal* but still *functional* version of this PR?
Or do things already work and this is strictly an optimization?

Reply:
Actually, no my code is correct. I'll add a comment as to why.